### PR TITLE
fix(formatter): incompatibility for class and implements 

### LINKED ISSUE #4143: 📝  Formatting incompatibility between biome and prettier for class and implements
### Environment information

```bash
CLI:
  Version:                      1.9.2
  Color support:                true

Platform:
  CPU Architecture:             x86_64
  OS:                           linux

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  NO_COLOR:                     unset
  TERM:                         "xterm-256color"
  JS_RUNTIME_VERSION:           "v22.8.0"
  JS_RUNTIME_NAME:              "node"
  NODE_PACKAGE_MANAGER:         "npm/10.8.3"

Biome Configuration:
  Status:                       Loaded successfully
  Formatter disabled:           false
  Linter disabled:              false
  Organize imports disabled:    true
  VCS disabled:                 true

Formatter:
  Format with errors:           false
  Indent style:                 Space
  Indent width:                 2
  Line ending:                  Lf
  Line width:                   80
  Attribute position:           Auto
  Bracket spacing:              BracketSpacing(true)
  Ignore:                       [".eslintrc.js", "**/assets/**", "front/**"]
  Include:                      ["**/*.ts"]

JavaScript Formatter:
  Enabled:                      false
  JSX quote style:              Double
  Quote properties:             AsNeeded
  Trailing commas:              All
  Semicolons:                   Always
  Arrow parentheses:            Always
  Bracket spacing:              BracketSpacing(true)
  Bracket same line:            false
  Quote style:                  Single
  Indent style:                 unset
  Indent width:                 unset
  Line ending:                  unset
  Line width:                   unset
  Attribute position:           Auto

JSON Formatter:
  Enabled:                      true
  Indent style:                 unset
  Indent width:                 unset
  Line ending:                  unset
  Line width:                   unset
  Trailing Commas:              unset

CSS Formatter:
  Enabled:                      true
  Indent style:                 unset
  Indent width:                 unset
  Line ending:                  unset
  Line width:                   unset
  Quote style:                  Double

GraphQL Formatter:
  Enabled:                      false
  Indent style:                 unset
  Indent width:                 unset
  Line ending:                  unset
  Line width:                   unset
  Bracket spacing:              unset
  Quote style:                  unset

Workspace:
  Open Documents:               0
```

### Configuration

```JSON
{
  "$schema": "https://biomejs.dev/schemas/1.9.2/schema.json",
  "vcs": { "enabled": false, "clientKind": "git", "useIgnoreFile": false },
  "files": { "ignoreUnknown": false, "ignore": [] },
  "formatter": {
    "enabled": true,
    "useEditorconfig": true,
    "formatWithErrors": false,
    "indentStyle": "space",
    "indentWidth": 2,
    "lineEnding": "lf",
    "lineWidth": 80,
    "attributePosition": "auto",
    "bracketSpacing": true,
    "include": ["**/*.ts"],
    "ignore": [".eslintrc.js", "**/assets/**", "front/**"]
  },
  "organizeImports": {
    "enabled": false
  },
  "linter": {
    "enabled": true,
    "rules": {
      "recommended": true,
      "style": {
        "useImportType": "off",
        "noParameterAssign": "off",
        "noNonNullAssertion": "off"
      },
      "suspicious": {
        "noExplicitAny": "off",
        "noAsyncPromiseExecutor": "off"
      },
      "complexity": {
        "noUselessConstructor": "off",
        "noThisInStatic": "off",
        "noStaticOnlyClass": "off",
        "noBannedTypes": "off"
      },
      "performance": { "noDelete": "off" }
    },
    "include": ["**/*.ts"],
    "ignore": [".eslintrc.js", "**/assets/**", "front/**", "**/test/**"]
  },
  "javascript": {
    "formatter": {
      "jsxQuoteStyle": "double",
      "quoteProperties": "asNeeded",
      "trailingCommas": "all",
      "semicolons": "always",
      "arrowParentheses": "always",
      "bracketSameLine": false,
      "quoteStyle": "single",
      "attributePosition": "auto",
      "bracketSpacing": true
    },
    "parser": {
      "unsafeParameterDecoratorsEnabled": true
    }
  }
}
```

### Playground link

https://biomejs.dev/playground/?indentStyle=space&code=ZQB4AHAAbwByAHQAIABjAGwAYQBzAHMAIABKAGkAcgBhAEMAcgBlAGEAdABlAFAAaQB4AEYAcgBhAHUAZABEAGUAdABlAGMAdABpAG8AbgBHAGEAdABlAHcAYQB5AAoAIAAgAGkAbQBwAGwAZQBtAGUAbgB0AHMACgAgACAAIAAgAFAAaQBjAGsAPABJAHMAcwB1AGUAUABpAHgARgByAGEAdQBkAEQAZQB0AGUAYwB0AGkAbwBuAEcAYQB0AGUAdwBhAHkALAAgACcAYwByAGUAYQB0AGUAUABpAHgARgByAGEAdQBkAEQAZQB0AGUAYwB0AGkAbwBuAEkAcwBzAHUAZQAnAD4ACgB7AAoAIAAgAGMAbwBuAHMAdAByAHUAYwB0AG8AcgAoAHAAcgBpAHYAYQB0AGUAIAByAGUAYQBkAG8AbgBsAHkAIABsAG8AZwBnAGUAcgA6ACAATABvAGcAZwBlAHIAKQAgAHsACgAgACAAIAAgAHQAaABpAHMALgBsAG8AZwBnAGUAcgAgAD0AIABsAG8AZwBnAGUAcgAuAGMAaABpAGwAZAAoAHsACgAgACAAIAAgACAAIABjAG8AbgB0AGUAeAB0ADoAIABKAGkAcgBhAEMAcgBlAGEAdABlAFAAaQB4AEYAcgBhAHUAZABEAGUAdABlAGMAdABpAG8AbgBHAGEAdABlAHcAYQB5AC4AbgBhAG0AZQAsAAoAIAAgACAAIAB9ACkAOwAKACAAIAB9AAoAfQA%3D

### Code of Conduct

- [X] I agree to follow Biome's Code of Conduct

### PR DESCRIPTION
**Pull request by mdm317**
fixes #4143 
## Summary
```
  "class",
  group("#1", [
    indent([
      " LongClassName",
      soft_line_break_or_space,
      "implements",
      group([
        indent([
          soft_line_break_or_space,
          "InterfaceNameLengthIsSixtySixCharacterssssssssssssssssssssssssssss"
        ])
      ])
    ])
  ]),
  " ",  <===== this
  if_group_breaks("#1", [hard_line_break]),
  "{",
  indent([hard_line_break, " constructer", group(["()"]), " {}"]),
  hard_line_break,
  "}",
```
When checking the operation of Prettier, the fits function calculates the width of an empty string as 1 instead of 0. The empty string refers to line 15 in the code, and we are using it as a space.
So i changed space to hardspace

## Test Plan

add test case

### CODE REVIEW

**Comment by ah-yu** [comment]
Why Prettier breaks `implements ...` even if the line doesn't exceed the width limit
image

**Comment by mdm317** [comment]
Thank you for asking.

I also thought it strange at first, so I checked how Prettier was working.
But, I'm not very familiar with the Prettier library, and I don't know whether this is what they intended or if it's a bug.

Prettier would sometimes break a line even when it hadn't exceeded the line length.
It seems like the example I worked on before was one of those cases. [previous pr](https://github.com/biomejs/biome/pull/1929)

Do you think this issue is a Prettier error rather than something we need to change?

**Comment by ah-yu** [comment]
> Do you think this issue is a Prettier error rather than something we need to change?

I personally tend to think this is a bug in Prettier.

**Comment by mdm317** [comment]
Thank you for sharing your thoughts! @ah-yu 

@dyc3 
Could you also share your thoughts?

**Comment by dyc3** [comment]
I also feel like this is a bug in prettier, but I also think we should generally try to match prettier. The fix here doesn't seem that complicated.

**Comment by Conaclos** [comment]
We could open an issue on the Prettier repository.

**Comment by dyc3** [comment]
I've opened https://github.com/prettier/prettier/issues/16723

File: crates/biome_js_formatter/tests/specs/ts/class/implements_clause.ts.snap [hunk 1]
```diff
@@ -8,6 +8,14 @@ info: ts/class/implements_clause.ts
 class ClassName implements Interface { }
 
 class LongClassName implements Interface1, Interface2, Interface3, Interface4, Interface5 { }
+
+class LongClassName implements InterfaceNameLengthIsSixtySixCharacterssssssssssssssssssssssssssss {
+    constructer(){}
```
Comment (branberry) [inline crates/biome_js_formatter/tests/specs/ts/class/implements_clause.ts.snap:13]: NIT: typo

**Comment by ematipico** [comment]
Closing, it's a prettier bug and they are fixing it

**Pull request closed without merging**